packages/workspace/src/browser/workspace-service.ts

```diff
@@ -82,84 +87,110 @@ export class WorkspaceService implements FrontendApplicationContribution {
         this.deferredRoots = new Deferred<FileStat[]>();
         this.deferredRoots.resolve(this._roots);
+        this.deferredAllRoots = new Deferred<FileStat[]>();
+        this.deferredAllRoots.resolve(this._allRoots);
         this.onWorkspaceChangeEmitter.fire(this._roots);
     }
 
-    protected async computeRoots(): Promise<FileStat[]> {
-        const config = await this.getWorkspaceConfig();
-        if (!config || !config.roots.length) {
-            return this._workspaceFolder ? [this._workspaceFolder] : [];
-        }
-        const roots: FileStat[] = [];
-        for (const rootUri of config.roots) {
-            const root = await this.toValidRoot(rootUri);
-            if (root) {
-                roots.push(root);
+    protected async computeRoots(workspaceFile: FileStat | undefined): Promise<{ found: FileStat[], all: FileStat[] }> {
+        const found: FileStat[] = [];
+        const all: FileStat[] = [];
+        if (workspaceFile) {
+            if (workspaceFile.isDirectory) {
+                return { found: [workspaceFile], all: [] };
             }
-        }
-        return roots;
-    }
 
-    protected async getWorkspaceConfig(): Promise<{ stat: FileStat, roots: string[] } | undefined> {
-        if (!this.preferences['workspace.supportMultiRootWorkspace']) {
-            return undefined;
-        }
-        const configUri = this.getWorkspaceConfigFileUri();
-        if (configUri) {
-            const uriStr = configUri.path.toString();
-            if (await this.fileSystem.exists(uriStr)) {
-                const { stat, content } = await this.fileSystem.resolveContent(uriStr);
-                if (content) {
-                    // FIXME use jsonc + ajx to parse & validate content
-                    const roots: string[] = JSON.parse(content).roots || [];
-                    return { stat, roots };
+            const workspaceData = await this.getWorkspaceDataFromFile(workspaceFile);
+            if (workspaceData) {
+                for (const { path } of workspaceData.folders) {
+                    const valid = await this.toValidRoot(path);
+                    if (valid) {
+                        found.push(valid);
+                        all.push(valid);
+                    } else {
+                        all.push({
+                            uri: path,
+                            lastModification: Date.now(),
+                            isDirectory: true
+                        });
+                    }
+                }
+                if (found.length < all.length) {
+                    const notFound = all.filter(a => found.findIndex(f => f.uri === a.uri) < 0)
+                        .map(r => r.uri)
+                        .join(', ');
+                    this.logger.error(`Unable to open root(s) ${notFound}`);
                 }
             }
         }
-        return undefined;
+        return { found, all };
```

Comment:
i find extensions are either interested in either `valid roots` or `all roots`. I can definitely return `{found, notFound}` - that way I would need to concatenate arrays every time I need `all roots`.

URIs can always be transformed into FileStats, and vice versa. The followings need FileStats:
- lableProvider.getName & lableProvider.getIcon need FileStats
- DirNode.createRoot needs FileStats
- FileSystem.getCurrentUserHome returns FileStat, so having FileStat exposed is easier for comparison
- dialogService needs FileStat to perform further operations
After static analysis I would say the usage of FileStats vs URIs is approx 40% and 60%. I prefer to do it in a separate PR to make both reviews easier. what do you think ?

Reply:
@svenefftinge WDY? I would like to explore an alternative then roots are not validated URIs and it is up to the client to get a file stat for them. It will simplify the workspace service and make sure that clients are using valid file stats, not something is cached by the workspace service.

Reply:
this is what @marcdumais-work suggested a few weeks ago:

maybe we could expose the following data structure instead of the FileStat[]:
```ts
interface Roots {
  stat: FileStat;
  existing: boolean;
  uri: URI
}
```
This way we only need to expose one array, and all the other theia extensions don't have to duplicate the logic of checking if the root exists or not.

Reply:
I think the service should just expose what is configured and doesn't take care of validating the URIs. Clients should take care of that. After all, folders can be deleted the moment they are exposed as valid.